registry/storage/garbagecollect_test.go

```diff
@@ -209,53 +209,71 @@ func TestDeleteManifestIfTagNotFound(t *testing.T) {
 	manifestService, _ := repo.Manifests(ctx)
 
 	// Create random layers
-	randomLayers1, err := testutil.CreateRandomLayers(3)
+	randomLayers, err := testutil.CreateRandomLayers(12)
```

Comment:
Rather than expanding this test, can we leave it as is and write new tests? I think we'd need new tests which ensure at least the following are covered:


- untagged manifests which are referenced by a tagged manifest list are kept when `RemoveUntagged:true`
    - blobs which are uniquely referenced by an untagged manifest which are referenced by a tagged manifest list are kept when `RemoveUntagged:true`
- untagged manifests which are referenced by an untagged manifest list are removed when `RemoveUntagged:true`
    - blobs which are uniquely referenced by an untagged manifest which are referenced by an untagged manifest list are removed when `RemoveUntagged:true`
- tagged manifests which are referenced by an untagged manifest list are kept when `RemoveUntagged:true`

Also, we should use schema 2 manifests and manifest lists in these tests. Schema 1 is very deprecated and isn't realistic in the scenario. `testutil` has functions for schema 2 manifests already, but not manifest lists.

Reply:
Also, we need to try to ensure that there are no ordering issues: ensuring that an untagged manifest referenced by a tagged manifest list is handled correctly regardless of whether it is encountered by the manifest enumerator before or after the manifest list which references it.

For example, we need to be sure to avoid deleting a manifest (or not marking its blobs, ect), only later to find it was referenced by another manifest list that was tagged.

Reply:
Agreed we should make those updates to the tests.

As far as ordering goes is there some easy way I can control what order manifests are enumerated?

Reply:
> As far as ordering goes is there some easy way I can control what order manifests are enumerated?

Not really :sweat_smile: 

I think as long as we provide several randomly created lists with randomly created manifests per test we'll have this covered, otherwise the test will flake, or maybe even only occasionally pass (reverse-flake?). It's not perfect, but it should get caught relatively soon.

To test for this every time always, you'd need to manually build out a list and manifests that happen to produce digests that lexicographically sort before and after the lists digests. It's a little extra to be honest, but it's fine if you chose to have a test like this or not.